frontend/src/components/AppState.js

```diff
@@ -3,76 +3,109 @@ import React from 'react';
 export const AppContext = React.createContext(null);
 
 export const initialState = {
-  payload: {
+  businessData: {
     assignments: [], submissions: [], students: [], denormalized_data: [],
   },
-  compareAssignments: {
+  visualizationData: {
     heatMapData: [],
-    pivoted: false,
+  },
+  processing: {
+    loadingBusinessData: true,
+    pivotingHeatMap: false,
+    pivotedHeatMap: false,
     error: false,
-    message: '',
+    errorMessage: '',
+  },
+  controls: {
     selectors: {
       showingRubrics: { values: ['All assignments'], selected: 'All assignments' },
       sortBy: { values: ['Due Date'], selected: 'Due Date' },
       sections: { values: ['All Sections'], selected: 'All Sections' },
       instructors: { values: ['All Instructors'], selected: 'All Instructors' },
     },
   },
-  loaded: false,
-  placeholder: 'Loading',
+  navigation: {
+    activeTab: 'compareAssignments',
+  },
+};
+
+// Helper to refactor the process of updating a selector's set of choices
+const updateSelectorValues = (selectorName, value, state) => {
+  const returnState = { ...state };
+  returnState.controls.selectors[selectorName].values = value;
+  return returnState;
+};
+
+// Helper to refactor the process of updating a selector's selection
+const updateSelectorSelection = (selectorName, value, state) => {
+  const returnState = { ...state };
+  returnState.controls.selectors[selectorName].selected = value;
+  return returnState;
 };
 
 export const reducer = (state, action) => {
   switch (action.type) {
-    case 'receivePayload':
+    case 'businessDataFetching':
```

Comment:
An interesting ongoing discussion you may be aware of with regards to selectors: https://github.com/reactjs/rfcs/pull/119

Reply:
A lot to read there, but I'll go over it. Thanks for sharing